src/core/nodeOps.test.ts

```diff
@@ -301,6 +301,37 @@ describe('nodeOps', () => {
       expect(node.defines[allCapsUnderscoresKey]).equals(allCapsUnderscoresValue)
     })
 
+    it('sets "on*" callbacks on Object3D', () => {
+      // Issue: https://github.com/Tresjs/tres/issues/360
+      const { createElement, patchProp } = nodeOps
+      const object = createElement('TresObject3D', undefined, undefined, {})
+
+      const onAfterRender = vi.fn()
+      const onAfterShadow = vi.fn()
+      const onBeforeRender = vi.fn()
+      const onBeforeShadow = vi.fn()
+
+      patchProp(object, 'onAfterRender', null, onAfterRender)
+      patchProp(object, 'onAfterShadow', null, onAfterShadow)
+      patchProp(object, 'onBeforeRender', null, onBeforeRender)
+      patchProp(object, 'onBeforeShadow', null, onBeforeShadow)
+
+      expect(onAfterRender).not.toBeCalled()
+      expect(onAfterShadow).not.toBeCalled()
+      expect(onBeforeRender).not.toBeCalled()
+      expect(onBeforeShadow).not.toBeCalled()
+
+      object.onAfterRender()
+      object.onAfterShadow()
+      object.onBeforeRender()
+      object.onBeforeShadow()
+
+      expect(onAfterRender).toBeCalled()
+      expect(onAfterShadow).toBeCalled()
+      expect(onBeforeRender).toBeCalled()
+      expect(onBeforeShadow).toBeCalled()
```

Comment:
Should we be testing that cbs were called

It looks like we'd want to test that the patchProp code you added sets the on* callbacks

At first glance it would be something like
```suggestion
      // on* cbs not set by default
      expect(object.onAfterRender).toEqual(undefined)
      expect(object.onAfterShadow).toEqual(undefined)
      expect(object.onBeforeRender).toEqual(undefined)
      expect(object.onBeforeShadow).toEqual(undefined)

      // patchProp should set on* cbs on object
      patchProp(object, 'onAfterRender', null, onAfterRender)
      patchProp(object, 'onAfterShadow', null, onAfterShadow)
      patchProp(object, 'onBeforeRender', null, onBeforeRender)
      patchProp(object, 'onBeforeShadow', null, onBeforeShadow)

      // Assert that on* cbs were set on object
      expect(object.onAfterRender).toEqual(onAfterRender)
      expect(object.onAfterShadow).toEqual(onAfterShadow)
      expect(object.onBeforeRender).toEqual(onBeforeRender)
      expect(object.onBeforeShadow).toEqual(onBeforeShadow)
```

Reply:
Hey @garrlker ,

Thanks for the comments.

```
      expect(object.onAfterRender).toEqual(undefined)
      expect(object.onAfterShadow).toEqual(undefined)
      expect(object.onBeforeRender).toEqual(undefined)
      expect(object.onBeforeShadow).toEqual(undefined)
```

This will always fail. These methods are defined by THREE during the constructor call:

https://github.com/mrdoob/three.js/blob/master/src/core/Object3D.js#L123

----

```
      // Assert that on* cbs were set on object
      expect(object.onAfterRender).toEqual(onAfterRender)
      expect(object.onAfterShadow).toEqual(onAfterShadow)
      expect(object.onBeforeRender).toEqual(onBeforeRender)
      expect(object.onBeforeShadow).toEqual(onBeforeShadow)
```

Sure thing. I'll update.